include/operations/blas2_trees.h

```diff
@@ -26,27 +26,44 @@
 #ifndef SYCL_BLAS_BLAS2_TREES_H
 #define SYCL_BLAS_BLAS2_TREES_H
 namespace blas {
-template <typename output_t, typename matrix_t, typename vector_t>
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          bool is_transposed, int cache_line_size, int work_per_thread>
 struct Gemv {
   using value_t = typename vector_t::value_t;
   using index_t = typename vector_t::index_t;
-  output_t lhs_;
-  matrix_t matrix_;
-  vector_t vector_;
+  lhs_t lhs_;
+  matrix_t matrix_a_;
+  vector_t vector_x_;
+  index_t wGs_per_NC_;
+  index_t wGs_per_C_;
+  index_t local_memory_size_;
 
-  Gemv(output_t &_l, matrix_t &_matrix, vector_t &_vector);
-  index_t get_size() const;
+  Gemv(lhs_t &_l, matrix_t &_matrix, vector_t &_vector, index_t &_WGs_per_row,
+       index_t &_WGs_per_col, index_t &_local_memory_size);
   bool valid_thread(cl::sycl::nd_item<1> ndItem) const;
-  value_t eval(index_t i);
   value_t eval(cl::sycl::nd_item<1> ndItem);
+  template <typename local_memory_t>
+  value_t eval(local_memory_t local_mem, cl::sycl::nd_item<1> ndItem);
   void bind(cl::sycl::handler &h);
   void adjust_access_displacement();
+
+ private:
+  template <typename ScratchPointerType>
+  void extract_input_block(ScratchPointerType scratch, const index_t &local_id,
+                           const index_t &group_id, const index_t &local_range,
+                           const index_t &lda, index_t mat_tile_id);
 };
 
-template <typename output_t, typename matrix_t, typename vector_t>
-Gemv<output_t, matrix_t, vector_t> make_gemv(output_t &lhs_, matrix_t &matrix_,
-                                             vector_t &vector_) {
-  return Gemv<output_t, matrix_t, vector_t>(lhs_, matrix_, vector_);
+template <bool is_transposed, int cache_line_size, int work_per_thread,
+          typename lhs_t, typename matrix_t, typename vector_t>
+Gemv<lhs_t, matrix_t, vector_t, is_transposed, cache_line_size, work_per_thread>
+make_Gemv(lhs_t &lhs_, matrix_t &matrix_, vector_t &vector_,
+          typename vector_t::index_t WGs_per_NC_,
```

Comment:
I think we do not use a capital letter for naming variable. Can you fix that @przemektmalon

Reply:
@mehdi-goli No capital letters at all, or just the first letter ? Like is `wGs_per_NC_` ok ?

Reply:
@mehdi-goli I've gone with `wgs_per_NC_` and same format for others, looks a bit cleaner I think.

Reply:
thanks, change the NC to nc as well

Reply:
and also if there is _C, change it to c. Thanks!